lib/Sema/CSBindings.cpp

```diff
@@ -1733,31 +1733,35 @@ void BindingSet::dump(llvm::raw_ostream &out, unsigned indent) const {
       break;
 
     case AllowedBindingKind::Subtypes:
-      out << "(subtypes of) ";
+      return "(subtypes of) " + type.getString(PO);
       break;
 
     case AllowedBindingKind::Supertypes:
-      out << "(supertypes of) ";
+      return "(supertypes of) " + type.getString(PO);
       break;
     }
     if (auto *literal = binding.getDefaultedLiteralProtocol())
-      out << "(default from " << literal->getName() << ") ";
-    out << type.getString(PO);
+      return "(default from " + literal->getName().str().str() + ") " +
+             type.getString(PO);
   };
 
   out << "[with possible bindings: ";
-  interleave(Bindings, printBinding, [&]() { out << "; "; });
+  std::vector<std::string> potentialBindings;
```

Comment:
Here is an alternative approach - instead of dealing with all the string concatenation you could introduce a new struct to encode a "kind" enum (supertypes, subtypes, default type of literal) and a type with a print method that accepts `raw_ostream` and indent. The struct could be declared right before the vector (you can also use `SmallVector` for this since the number of bindings is usually pretty small).